bgpd/bgp_snmp.c

```diff
@@ -154,6 +156,196 @@ static int bgp_snmp_module_init(void)
 	return 0;
 }
 
+static struct peer *bgp_snmp_lookup_peer_in_vrf(struct bgp *bgp, const struct ipaddr *addr)
+{
+	struct peer *peer;
+	struct listnode *node;
+	sa_family_t peer_family;
+
+	if (!bgp)
+		return NULL;
+	/*
+	 * Iterate through the peers in the VRF
+	 */
+	if (!bgp->peer_by_addr)
+		return NULL;
+
+	/*
+	 * peer_by_addr is kept sorted by address, so walk it looking for the
+	 * first entry strictly greater than the supplied key. This works even if
+	 * the exact peer has disappeared between requests.
+	 */
+	for (ALL_LIST_ELEMENTS_RO(bgp->peer_by_addr, node, peer)) {
+		if (!peer->connection)
+			continue;
+
+		peer_family = sockunion_family(&peer->connection->su);
+
+		/* Check if the address matches */
+		if (peer_family != addr->ipa_type)
+			continue;
+
+		switch (peer_family) {
+		case AF_INET:
+			if (IPV4_ADDR_SAME(&peer->connection->su.sin.sin_addr, &addr->ipaddr_v4))
+				return peer;
+			break;
+		case AF_INET6:
+			if (IPV6_ADDR_SAME(&peer->connection->su.sin6.sin6_addr, &addr->ipaddr_v6))
+				return peer;
+			break;
+		default:
+			break;
+		}
+	}
+
+	return NULL;
+}
+
+static struct peer *bgp_snmp_get_first_peer_in_vrf(struct bgp *bgp, sa_family_t family)
+{
+	struct peer *peer;
+	struct peer *first_peer = NULL;
+	struct listnode *node;
+	struct listnode *head;
+
+	if (!bgp || !bgp->peer_by_addr)
```

Comment:
again this function is called in several places where the bgp pointer is already checked or we know it cannot be null.  Let's pick one place to do it and stick with that